### ISSUE TITLE
comparing short and full versions

### ISSUE THREAD

**Issue by omgbebebe** [fosskers/versions#7]
I have issue with comparing short and full notations of same versions. E.g:
```
> import Data.Versions
> Right v1 = parseV "1.2.0"
> Right v2 = parseV "1.2"
> compare v1 v2
GT
```
I think they should be equal. Or it's expected behavior?

**Comment by fosskers** [comment]
That's expected behaviour, actually. `v1` would parse as a `SemVer`, but `v2` would parse as a `Version`. To compare them, the `SemVer` is "down cast" to a Version, which has different comparison semantics. 

The "why" in this case is that the PVP versioning scheme (used by Haskell packages) makes this distinction: `0.2 < 0.2.0 < 0.2.0.0`

Did this come up as a "bug" in some code you were writing?

**Comment by omgbebebe** [comment]
I'm trying to parse a packages repository which have a really messy versioning schemes. I found this behavior strange when had try to understand why I get different results from various methods of finding maximum version. I made a gitst that demostrates my main problem. https://gist.github.com/omgbebebe/5776653e3d5ab43562aa155990e7db1c Maybe you could point me to my errors.

**Comment by fosskers** [comment]
Removing the first `v` in all of those versions should help you quite a bit. 

```haskell
λ parseV "v0.52"
Right (General (Version {_vEpoch = Nothing, _vChunks = [[Str "v",Digits 0],[Digits 52]], _vRel = []}))

λ parseV "0.0.34"
Right (Ideal (SemVer {_svMajor = 0, _svMinor = 0, _svPatch = 34, _svPreRel = [], _svMeta = []}))

λ parseV "v0.00000-testKSP105_5"
Right (Complex (VNode ["v0","00000"] VHyphen (VNode ["testKSP105"] VUnder (VLeaf ["5"]))))

λ (>) <$> parseV "v0.52" <*> parseV "v0.53"
Right False
```
Let me test a bit more.

**Comment by fosskers** [comment]
The two `Mess`es in your list *might* be throwing things off. Here's a disclaimer from the docs:

> [Messes are] Not guaranteed to have well-defined ordering (Ord) behaviour, but so far internal tests show consistency.

**Comment by fosskers** [comment]
Ok, this seems expected:
```haskell
λ map prettyV sorted
["v0.00000-testKSP105_4","v0.00000-testKSP105_5","v0.02","v0.04","v0.05","v0.06","v0.07","v0.08","v0.09","v0.10","v0.11","v0.12",
"v0.13","v0.14","v0.15","v0.16","v0.17","v0.18","v0.19","v0.20","v0.21","v0.22","v0.23","v0.24",
"v0.25","v0.26","v0.27","v0.28","v0.29","v0.30","v0.31","v0.32","v0.33","v0.35","v0.36","v0.37",
"v0.38","v0.39","v0.40","v0.41","v0.42","v0.43","v0.44","v0.45","v0.46","v0.47","v0.48","v0.49",
"v0.50","v0.51","v0.52","v0.53","0.0.34"]

**Comment by fosskers** [comment]
I think I figured it out. 
```haskell
"v0.00000-testKSP105_4","v0.00000-testKSP105_5","v0.02","v0.04","v0.05","v0.06","v0.07","v0.08","v0.09"
```
^ these all parse as `Mess`, and so they've clumped together "in order" here at the beginning of the list.  `v0.09` looks like a `Version` but it isn't because of the `09`. That's not a legal `Int`. In the original unsorted list, all these numbers are mixed together, so `Version` and `Mess` are being compared, and it's a coincidence of the ordering that the `v0.53` is being kicked out of the maximum spot and we see `v0.51` at the end.

I also switched from `sortBy` to `sort`, and from `maximumBy` to `maximum` in my tests. Also by removing the initial `v`, we see this order:
```haskell
λ main
0.53 :plain maximum
0.53 :last from sorted list
0.53 :maximum of sorted list

λ map prettyV sorted
["0.0.34","0.00000-testKSP105_4","0.00000-testKSP105_5","0.02","0.04","0.05","0.06","0.07","0.08",
"0.09","0.10","0.11","0.12","0.13","0.14","0.15","0.16","0.17","0.18","0.19","0.20","0.21","0.22","0.23",
"0.24","0.25","0.26","0.27","0.28","0.29","0.30","0.31","0.32","0.33","0.35","0.36","0.37","0.38","0.39",
"0.40","0.41","0.42","0.43","0.44","0.45","0.46","0.47","0.48","0.49","0.50","0.51","0.52","0.53"]
```

**Comment by omgbebebe** [comment]
Thanks for your patience!:) I really do not expect that the list with different version schemes will be on right order. The main question is why I get different results when I try to find "maximum" version with different approaches. [Another example without messy versions in list](https://gist.github.com/omgbebebe/b3be7483d91b5a714a4866734739aa89).
- `maximumBy compare versions` -> 0.0.34
- `last $ sort versions` -> v0.08
- `maximumBy compare $ sort versions` -> v0.48

In other words, `maximum versions` not always equal to `last (sort versions)` in case the list contains `Ideal` and `General` schemes.
Looks like that I missed something but I can't figureout what exacltly.:)

**Comment by omgbebebe** [comment]
If you're interested I've made a [`quickcheck` for this case](https://gist.github.com/omgbebebe/b47c11c0080e1e8ef7f2e5b12b5cb363).
```
> test2
*** Failed! Falsifiable (after 9 tests):  
["v0.24","0.0.34","v0.53","v0.42","v0.24","v0.53","v0.53","v0.05"]
```
Seems like you are right about leading zeros in versions. If I remove them from source list the tests are passed.

**Comment by fosskers** [comment]
Removing the initial `v` will also help you.

**Comment by omgbebebe** [comment]
Yes. Withot leading `v` all tests are passed. Thank you for help!

**Comment by fosskers** [comment]
My pleasure!
